✨ Is2392/api project version control

pcrespov opened the pull request:
<!-- Common title prefixes/annotations:

WIP: work in progress

Consider prefix your PR message with an emoticon
  🐛 bugfix
  ✨ new feature
  ♻️ refactoring
  💄 updates UI or 🚸 UX/usability
  🚑️ hotfix
  ⚗️ experimental
  ⬆️ upgrades dependencies
  📝 documentation
or from https://gitmoji.dev/

and append (⚠️ devops) if changes in devops configuration required before deploying
-->

## What do these changes do?

New projects repository concept: a git-like version control system for projects. It create snapshots of a project at a given checkpoint.

⚗️ NOTE: this feature is still under development and is only active with ``WEBSERVER_DEV_FEATURES_ENABLED=1 `` (e.g. in master deploy)

### Changes in ``webserver`` service
- ADDED ``repos`` app module in the ``webserver`` implemented in ``simcore_service_webserver/meta*.py`` files.
- CHANGES ``webserver`` **API**
    - new ``/repos/projects/*`` entrypoints
![image](https://user-images.githubusercontent.com/32402063/132520830-9acd96e0-d2e9-459a-a973-030bda825aa6.png)
   - keeps ``/projects/snapshot/*`` entrypoints until fully replaced by the former
   - ``sandbox/projects_openapi_generator.py`` prototypes API (i.e. only for dev purposes)
   
### Changes in ``postgres-database`` library
- CHANGES in ``postgres-database`` **tables**:
![Capture 2021-09-15 11-54-41](https://user-images.githubusercontent.com/32402063/133412598-545801fe-9a46-4306-b092-75521123e435.png) 
     -  ``projects_snaphots`` temporarily kept until fully replaced
     - NEW ``projects_vc_*`` tables to handle version control of projects
- ADDED **``simcore_postgres_database.utlis_aiopg_orm.BaseOrm``** support for pagination ``BaseOrm.fetch_page``
- 


## Related issue/s

- Builds on top of PR #2519
- Implementing #2392 : version control on projects


## How to test

### Automatic
```command line
make devenv
source .venv/bin/activate
cd services/web/server
make install-dev
pytest --pdb --ff tests/unit/with_dbs/10/meta
```

### Manual

- build and deploy
```command line
make .env
echo WEBSERVER_DEV_FEATURES_ENABLED=1 >>.env
make build-devel
make up-devel
```
- http://127.0.0.1.nip.io:9081 create user 
- http://127.0.0.1.nip.io:9081/dev/doc#/repository

## Checklist


- [x] Did you change any service's API? Then make sure to bundle document and upgrade version (``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Unit tests for the changes exist
- [ ] Runs in the swarm
- [x] Documentation reflects the changes
- [x] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)
- [x] Migration scripts

sanderegg commented on services/web/server/src/simcore_service_webserver/meta_models_repos.py:
```diff
@@ -0,0 +1,84 @@
+    class Config:
+        orm_mode = True
+
+    workbench: Dict[UUID, Node] = {}
```
should it not use [NodeID, Node]?

sanderegg commented on services/web/server/src/simcore_service_webserver/meta_models_repos.py:
```diff
@@ -0,0 +1,84 @@
+        orm_mode = True
+
+    workbench: Dict[UUID, Node] = {}
+    ui: Dict[UUID, Any] = {}
```
same here

pcrespov replied:
iti s the same, right? :-D

sanderegg replied:
yes it is the same. but aliases are there in the case they would change and to make it more understandable. Here you actually *know* these are the same. but for someone not knowing the code is a bit more difficult.

sanderegg commented on services/web/server/src/simcore_service_webserver/meta_models_repos.py:
```diff
@@ -0,0 +1,84 @@
+
+
+class RepoApiModel(BaseModel):
+    project_uuid: UUID
```
use ProjectID

GitHK commented on packages/service-library/src/servicelib/aiohttp/rest_responses.py:
```diff
@@ -87,37 +87,40 @@ def create_data_response(
     """
+    if not isinstance(errors, list):
+        errors = [
+            errors,
```
could you please remove the comma and reformat?

sanderegg commented on services/web/server/src/simcore_service_webserver/projects/projects_handlers.py:
```diff
@@ -273,7 +274,7 @@ async def replace_project(request: web.Request):
     user_id: int = request[RQT_USERID_KEY]
     try:
-        project_uuid = request.match_info["project_id"]
+        project_uuid = UUID(request.match_info["project_id"])
```
use ProjectID

sanderegg commented on services/web/server/src/simcore_service_webserver/projects/projects_handlers.py:
```diff
@@ -290,7 +291,7 @@ async def replace_project(request: web.Request):
         context={
             "dbapi": db,
-            "project_id": project_uuid,
+            "project_id": str(project_uuid),
```
very minor: it seems doing f"{}" is more efficient than using str()

sanderegg commented on services/web/server/tests/sandbox/projects_openapi_generator.py:
```diff
@@ -55,10 +138,39 @@ class Node(BaseModel):
 class Project(BaseModel):
+    """Domain model"""
+
     id: UUID
```
user ProjectID

sanderegg commented on services/web/server/tests/sandbox/projects_openapi_generator.py:
```diff
@@ -55,10 +138,39 @@ class Node(BaseModel):
+    """Domain model"""
+
     id: UUID
     pipeline: Dict[UUID, Node]
```
use NodeID

pcrespov replied:
these are not that important since I only use them for prototyping

sanderegg replied:
yes I know... but expecting they come in one day

sanderegg commented on services/web/server/tests/sandbox/projects_openapi_generator.py:
```diff
@@ -55,10 +138,39 @@ class Node(BaseModel):
+
+# requests models
+class ProjectNew(BaseModel):
+    pipeline: Dict[UUID, Node]
```
same

sanderegg commented on services/web/server/tests/sandbox/projects_openapi_generator.py:
```diff
@@ -55,10 +138,39 @@ class Node(BaseModel):
+# response models
+class ProjectItem(BaseModel):
+    # Lightweight and part of an array
+    id: UUID
```
same

sanderegg commented on services/web/server/tests/sandbox/projects_openapi_generator.py:
```diff
@@ -55,10 +138,39 @@ class Node(BaseModel):
+
+
+class ProjectDetail(BaseModel):
+    id: UUID
```
same and be careful with this name. cause I think in the database we actually have both id and uuid, the first one being an integer and the second a uuid

sanderegg commented on services/web/server/tests/unit/with_dbs/10/meta/conftest.py:
```diff
@@ -0,0 +1,189 @@
+
+
+@pytest.fixture()
+def project_uuid(user_project: ProjectDict) -> UUID:
```
use ProjectID

sanderegg commented on services/web/server/tests/unit/with_dbs/10/meta/test_meta_api_handlers_repos.py:
```diff
@@ -0,0 +1,228 @@
+    resp = await client.get(f"/{vtag}/projects/{project_uuid}")
+    data, _ = await assert_status(resp, web.HTTPOk)
+    project = Project.parse_obj(data)
+    assert project.uuid == UUID(project_uuid)
```
do you actually need this casting here? if yes we might have a problem

pcrespov replied:
@sanderegg the left is a UUID and right one is a string (coming from a faker)
```
  assert project.uuid == project_uuid
E       AssertionError: assert equals failed
E         UUID('4bd91037-8591-4c5c-a4e9-b3fe1476ca0b')  '4bd91037-8591-4c5c-a4e9-b3fe1476ca0b'
```

sanderegg commented on services/web/server/tests/unit/with_dbs/10/meta/test_meta_api_handlers_repos.py:
```diff
@@ -0,0 +1,228 @@
+    checkpoint2 = CheckpointApiModel.parse_obj(data)  # NOTE: this is NOT API model
+
+    assert (
+        checkpoint1 == checkpoint2
```
nice!

sanderegg commented on services/web/server/tests/unit/with_dbs/10/meta/test_meta_core_repos.py:
```diff
@@ -0,0 +1,115 @@
+# TESTS
+
+
+@pytest.mark.acceptance_test
```
what is the purpose with these acceptance_test(s)? I see that you mark these but what are you doing with them?

pcrespov replied:
it is a small experiment ... it is something i am reading in "robust python"

sanderegg reviewed (changes requested):
Very nice and clean. I think we should in general use the ProjectID/NodeID aliases instead of putting UUID everywhere and str at some locations.
Also I am not sure I understand why the whole is called meta? that is for meta-modeling of course, but the function here is about version control no? meta is a very wide term.
a few small things, maybe worth to check.

GitHK commented on services/web/server/src/simcore_service_webserver/meta_api_handlers_base.py:
```diff
@@ -0,0 +1,105 @@
+    return url_for
+
+
+# FIXME: access rights using same approach as in access_layer.py in storage.
```
I guess this is for future developments

GitHK commented on services/web/server/src/simcore_service_webserver/meta_api_handlers_repos.py:
```diff
@@ -0,0 +1,323 @@
+    url_for = create_url_for_function(request)
+    vc_repo = VersionControlRepository(request)
+
+    _limit = int(request.query.get("limit", 20))
```
should we not have a shared constant DEFAULT_ITEMS_PER_PAGE across the entire app?
@sanderegg I don't remember if you already added something like this.

pcrespov replied:
will add in ``packages/service-library/src/servicelib/rest_pagination_utils.py``

GitHK commented on services/web/server/src/simcore_service_webserver/meta_api_handlers_repos.py:
```diff
@@ -0,0 +1,323 @@
+    vc_repo = VersionControlRepository(request)
+
+    _project_uuid = request.match_info["project_uuid"]
+    _limit = int(request.query.get("limit", 20))
```
also use the same constant here

GitHK commented on services/web/server/src/simcore_service_webserver/meta_api_handlers_snapshots.py:
```diff
@@ -312,37 +232,9 @@ async def _update_snapshot(
-    )
-
-    return params
+rename_routes_as_handler_function(routes, prefix=__name__)
```
are the names of the handles here importa? will renaming cause issues? If yes, also add a comment here as well.

pcrespov replied:
yes, the routes name are very important. Used to produce reverse urls or to test the API implementation vs the OAS

GitHK commented on services/web/server/src/simcore_service_webserver/meta_core_repos.py:
```diff
@@ -0,0 +1,155 @@
+    return WorkbenchView.from_orm(content)
+
+
+list_repos_safe = validate_arguments(list_repos, config=CFG)
```
I guess this was not applied as a decorator because you needed both versions?

pcrespov replied:
yes

GitHK commented on services/web/server/src/simcore_service_webserver/meta_db.py:
```diff
@@ -0,0 +1,479 @@
+log = logging.getLogger(__name__)
+
+
+def eval_checksum(workbench: Dict[str, Any]) -> SHA1Str:
```
May I ask you to remove the eval from the name, I perceive it as a security concern. Maybe `get_checksum` or `compute_checksum`.

pcrespov replied:
security concern?
Perhaps we should start to make a list as in [here](https://melevir.medium.com/python-functions-naming-the-algorithm-74320a18278d)

GitHK commented on services/web/server/src/simcore_service_webserver/meta_db.py:
```diff
@@ -0,0 +1,479 @@
+            super().__init__(
+                projects_vc_snapshots,
+                connection,
+                writeonce={"checksum"},  # TODO:  all? cannot delete snapshots?
```
is this relevant?

GitHK commented on services/web/server/src/simcore_service_webserver/meta_db.py:
```diff
@@ -0,0 +1,479 @@
+    async def init_repo(self, project_uuid: UUID) -> int:
+        async with self.engine.acquire() as conn:
+
+            async with conn.begin():
```
Since you are editing multiple tables, should this not be under a transaction? In case of errors you can rollback and return raise the error again.

pcrespov replied:
that is indeed with ``conn.begin()`` does, right? SEE https://aiopg.readthedocs.io/en/stable/sa.html#aiopg.sa.SAConnection.begin

GitHK commented on services/web/server/src/simcore_service_webserver/meta_db.py:
```diff
@@ -0,0 +1,479 @@
+
+                return repo_id
+
+    async def commit(
```
same concern as above, please consider using a transaction

pcrespov replied:
same as above. I already use  ``async with conn.begin()`` as well

GitHK commented on services/web/server/src/simcore_service_webserver/meta_db.py:
```diff
@@ -0,0 +1,479 @@
+
+            return logs, total_count
+
+    async def update_annotations(
```
same concern as above, please consider using a transaction

pcrespov replied:
thisone i forgot :-) thx

GitHK commented on services/web/server/src/simcore_service_webserver/meta_db.py:
```diff
@@ -0,0 +1,479 @@
+
+            return repo.id, commit_id
+
+    async def checkout(self, repo_id: int, commit_id: int) -> int:
```
same concern as above, please consider using a transaction

pcrespov replied:
same as above.

GitHK commented on services/web/server/tests/unit/with_dbs/10/meta/test_meta_core_repos.py:
```diff
@@ -0,0 +1,115 @@
+    assert checkpoint1.tags == ("v0",)
+    assert checkpoint1.message == "first commit"
+
+    # TODO: project w/o changes, raise error .. or add new tag?
```
is this still valid?

GitHK reviewed (commented):
Very Nice. Some concerns and questions below.

pcrespov commented on packages/postgres-database/src/simcore_postgres_database/utils_aiopg_orm.py:
```diff
@@ -156,35 +156,80 @@ async def fetch(
+    ) -> Tuple[List[RowProxy], int]:
+        """Support for paginated fetchall
+
+        IMPORTANT: pages are sorted by primary-key
```
in the case of paging it is not an option ... it is about consistency, right? you need an order.

GitHK replied:
Since the data will always be listed in the same way if there are no new elements added to the set, I think the default could be omitted. In theory it should use the primary for indexing. Omitting the sort will skip some steps form the internal query optimiser.

pcrespov replied:
based on the [tests i have performed](https://github.com/ITISFoundation/osparc-simcore/blob/46be7bcacb2af46bd2c1436dd37ee3bf5194e712/packages/postgres-database/tests/test_utils_aiopg_orm.py), my guess is that the order is the insertion order. Normally the primary key is an auto-incremented index and therefore has no effect, but if the primary key is e.g. a unique name, then the order is NOT folloing the primary key.

pcrespov commented on services/web/server/src/simcore_service_webserver/meta_api_handlers_base.py:
```diff
@@ -0,0 +1,105 @@
+    )
+
+
+HandlerCoroutine = Callable[[web.Request], Awaitable[Optional[web.Response]]]
```
this is only understood by mypy, right?
Regarding the handler above, I will use the annotation provided by ``from aiohttp.web_middlewares import _Handler
``

sanderegg replied:
you basically write this ```reveal_type(handler)``` then run mypy... and in the logs you see what it spit out.

pcrespov commented on services/web/server/src/simcore_service_webserver/meta_api_handlers_repos.py:
```diff
@@ -0,0 +1,323 @@
+@routes.get(f"/{vtag}/repos/projects")
+@login_required
+@permission_required("project.read")
+@handle_request_errors
```
the problem with the middleware is that it applies to *all* requests and we would have to add some filters at runtime. here is explicitly filter at "codetime" :-)

pcrespov commented on services/web/server/src/simcore_service_webserver/application_config.py:
```diff
@@ -94,7 +94,7 @@ def create_schema() -> T.Dict:
             addon_section("studies_dispatcher", optional=True): minimal_addon_schema(),
             addon_section("exporter", optional=True): minimal_addon_schema(),
-            addon_section("snapshots", optional=True): minimal_addon_schema(),
+            addon_section("meta", optional=True): minimal_addon_schema(),
```
i wanted to keep everything under a single app module but ok, i can split in several ... makes sense :-)

pcrespov commented on services/web/server/src/simcore_service_webserver/meta_errors.py:
```diff
@@ -0,0 +1,28 @@
+from pydantic.errors import PydanticErrorMixin
+
+
+class BaseMetaValueError(PydanticErrorMixin, ValueError):
```
I learned/copied it from pydantic!

pcrespov commented on packages/postgres-database/src/simcore_postgres_database/utils_aiopg_orm.py:
```diff
@@ -156,35 +156,80 @@ async def fetch(
+        assert total_count >= 0  # nosec
+
+        rows: List[RowProxy] = await result.fetchall()
+        return rows, total_count
```
enterable?utterable? ... you mean iterable? :-) What advantage would that bring? It is already a paginated (i.e. bound) list.

Regarding the ``total_count`` i agree that a separated call would be better. will add.

GitHK replied:
The usage would be, when you require extra processing before sending the data elsewhere. You only go through it once. Well that's my experience with other ORMs.

pcrespov replied:
I think i undertand what you mean but those are not only iterables involved ...
in any case, as I mention in the header this is a "small and convenient" approach for a very specific case
it is far from being a generic tool

GitHK commented on docs/releasing-workflow-instructions.md:
````diff
@@ -130,20 +130,24 @@ A bug was found in version 1.2.0 of the simcore stack. The team decides to fix i
-  make release-hotfix version=MAJ.MIN.PATCH (git_sha=OPTIONAL)
-  ```
+```bash
+git clone https://github.com/ITISFoundation/osparc-simcore.git
````
Also please mention that you need to have push access to the osparc repo

pcrespov replied:
if you do not have push access, you probably are not allowed to do releases :-)

GitHK replied:
No I can create releases but I cannot create a hotfix. This was relative to the hotfixes, right?

GitHK reviewed (approved):
👌

pcrespov commented on packages/postgres-database/tests/test_utils_aiopg_orm.py:
```diff
@@ -130,6 +130,44 @@ async def test_orm_fetch_defaults(
+    assert page2[0].rrid == "RRID:foo"
+
+    # same as fetchall if offset=0 and limit is None
+    page, total_rows = await scicrunch_orm.fetch_page("rrid", offset=0)
```
i forgot .. thx

pcrespov commented:
@GitHK [Last commit](https://github.com/pcrespov/osparc-simcore/commit/ad803a0039cce494f8870d887538daa2439f4ecf) fixes [the issue](https://github.com/ITISFoundation/osparc-simcore/runs/3634404838?check_suite_focus=true#step:9:281) with the ``dir_with_random_content`` fixture  that we discussed.

The problem was that in the for loop you iterated using a generator that would re-evaluate a *recursive* glob on the folder you were operating at every iteration ... this would lead to an infinite recursion that lead to such long files. Changing ``rglob("*")`` (with R) by ``glob("*")`` solved the problem. I guess there is also a random component but i had to move on ...

I also replaced the randomizer by faker which creates nice names without collisions

I also removed copies of this fixture that were added in another test

Please double check specifically that [commit](https://github.com/pcrespov/osparc-simcore/commit/ad803a0039cce494f8870d887538daa2439f4ecf) in case i missed something. Thx!
